Add IndexedDB persistence for React Query

jsnajdr opened the pull request:
Adds IndexedDB persistence to React Query, using their [experimental `persistQueryClient` plugin](https://github.com/tannerlinsley/react-query/blob/master/docs/src/pages/plugins/persistQueryClient.md).

What I had to do is:
- move creation of the `QueryClient` object to the boot code, the same place where the Redux store is created. The reason is that after creating the client, we need to `await` the persistence initialization, and the boot code is the best place for that.
- put the `queryClient` into the page.js `context.queryClient` attribute and the inject it into React layouts from there
- extract the code that loads state from IndexedDB and stores it in a memory object to a separate module, `client/state/persisted-state`, because it's now used outside Redux, too.

Generally, the lifecycle of the Redux store and the React Query client should now be exactly the same. Where they are created, initialized by persisted state, passed in page.js and React context... They always travel together.

**How to test:**
Go to a route that uses React Query, e.g., `/people/team`, and verify that a `query-state-XXX` key has been stored to IndexedDB, containing React Query Cache data. It's there alongside with all the existing `redux-state-XXX:YYY` keys for modular Redux reducers.

jsnajdr committed:
Add IndexedDB persistence for React Query

sgomes commented on client/boot/common.js:
```diff
@@ -408,44 +422,63 @@ const setupMiddlewares = ( currentUser, reduxStore ) => {
+}
 
-	debug( 'Main layout rendered.' );
+async function createQueryClient( userId ) {
```
Should this method perhaps live under `state`? Its equivalent in Redux land would be something like `createReduxStore`, if I'm understanding this correctly, which does live under `state`.

It just seems a bit too low-level for this module, with things like manual throttling and string building for the property names.

sgomes commented on client/state/initial-state.js:
```diff
@@ -230,7 +208,7 @@ function getInitialPersistedState( initialReducer, currentUserId ) {
 // `loadAllState` must have completed first.
```
These comments need to be updated to `loadPersistedState`.

sgomes reviewed (commented):
Just leaving a few comments, not an exhaustive review. This generally looks good; it's an interesting complement to Redux persistence that doesn't take a huge toll 👍

One thing that concerns me a bit is the 6KB increase to the login entry point. I imagine it wasn't loading `react-query` but now it has to? 😕

zaguiini commented:
This works great! Excited to have it!

What if I don't want to cache some information? Is there a way to deny certain `useQuery` calls from entering the cache?

jsnajdr committed:
Update references to loadAllState in comments

jsnajdr committed:
Extract createQueryClient to a module

jsnajdr committed:
Remove QueryClient from login entrypoint

jsnajdr commented:
Thanks for review @sgomes, I addressed all your comments:
- removed the `QueryClient` from the login entrypoint because it doesn't use it. The login entrypoint has its own `ReduxWrappedLayout` component that doesn't mount a `QueryClientProvider` at all.
- extracted `createQueryClient` to its own module, and making that module TypeScript from the beginning.
- updated the `loadAllState` comments.

flootr commented on client/state/query-client.ts:
```diff
@@ -0,0 +1,29 @@
+import { getPersistedStateItem, storePersistedStateItem } from 'calypso/state/persisted-state';
+
+export async function createQueryClient( userId: number | undefined ): Promise< QueryClient > {
+	const queryClient = new QueryClient();
```
I think we may need to set `cacheTime` to at least `MAX_AGE` for persist to work properly as documented in https://react-query.tanstack.com/plugins/persistQueryClient#usage.

I tried it by setting `cacheTime` to only a few seconds (default is 5 minutes). Went to `/team` and then `/followers`. Checked both queries are cached and after set cache time was due reloaded the browser window. Then the team query was missing from IndexedDB. (It may be you don't even need to reload the browser window)

Update: this is ofc only valid given we want to persist the cache for `MAX_AGE` as set in a few lines below.

jsnajdr replied:
That's a good point! I pushed a commit that configures the `cacheTime` as equal to the persistor `maxAge`. Otherwise, persisted data that were just restored from local storage would be treated as expired, because they are too old.

flootr commented on client/state/query-client.ts:
```diff
@@ -0,0 +1,29 @@
+	const queryClient = new QueryClient();
+	if ( shouldPersist() ) {
+		const storeKey = `query-state-${ userId ?? 'logged-out' }`;
+		const persistor = {
```
Is there a reason we don't use the provided [`createAsyncStoragePersistor`](https://react-query.tanstack.com/plugins/createAsyncStoragePersistor) ([source](https://github.com/tannerlinsley/react-query/blob/master/src/createAsyncStoragePersistor-experimental/index.ts))? Is it because of our combined usage with redux persistance?

jsnajdr replied:
Yes, the `createAsyncStoragePersistor` didn't look like a good fit to me. Our "read" operation is always sync, because the state is preloaded at boot, we don't want to `JSON.stringify` on write because IndexedDB can store the cache object itself without serializing it... Instead of implementing a `Persistor` interface I would have to implement the almost identical `AsyncStorage` one, and the only value-added by `AsyncStoragePersistor` is throttling the "write" operation?

flootr replied:
Thanks for elaborating, makes sense 👍🏻

jsnajdr committed:
Set cacheTime equal to persistence maxAge

jsnajdr commented:
> What if I don't want to cache some information? Is there a way to deny certain `useQuery` calls from entering the cache?

Do you mean storing the data in the memory `QueryCache` object? That can be achieved by setting expire time to `0` or something like that.

Or preventing certain queries from being persisted? That would be nice and our Redux store can do that -- persistence is opt-in -- but I don't know at this moment how to do it with React Query. This PR persists everything.

zaguiini commented:
> Or preventing certain queries from being persisted? That would be nice and our Redux store can do that -- persistence is opt-in -- but I don't know at this moment how to do it with React Query. This PR persists everything.

That 👆🏻 I think a mechanism to prevent certain queries of being persisted would be beneficial.

zaguiini commented:
@jsnajdr,

Looking at `react-query`'s documentation, that's the closest I could get: https://codesandbox.io/s/pedantic-agnesi-71dmf

I had to hack into query key since `react-query`, unlike `apollo-client`, does not support metadata on queries. I will try and file a PR on `react-query`'s repo so we can have that and stop hacking their library 😛

flootr reviewed (approved):
LGTM 🚢
